src/system/tests/TestSystemTimer.cpp

```diff
@@ -206,6 +207,72 @@ static void CheckStarvation(nlTestSuite * inSuite, void * aContext)
     ServiceEvents(lSys, sleepTime);
 }
 
+void HandleStartTimer(Layer * inetLayer, void * aState, Error aError)
+{
+    TestContext & lContext = *static_cast<TestContext *>(aState);
+    NL_TEST_ASSERT(lContext.mTestSuite, true);
+    sStartTimerHandled = true;
+}
+
+static void CheckStartTimer(nlTestSuite * inSuite, void * aContext)
+{
+    uint32_t timeout_100ms = 100;
+
+    TestContext & lContext = *static_cast<TestContext *>(aContext);
+    Layer & lSys           = *lContext.mLayer;
+
+    sStartTimerHandled = false;
+
+    Error err = lSys.StartTimer(timeout_100ms, HandleStartTimer, aContext);
+    NL_TEST_ASSERT(lContext.mTestSuite, err == CHIP_SYSTEM_NO_ERROR);
+
+    while (!sStartTimerHandled)
+    {
+        struct timeval sleepTime;
+        sleepTime.tv_sec  = 0;
+        sleepTime.tv_usec = 1000; // 1 ms tick
+        ServiceEvents(lSys, sleepTime);
+    }
+
+    NL_TEST_ASSERT(lContext.mTestSuite, sStartTimerHandled == true);
+    lSys.CancelTimer(HandleStartTimer, aContext);
+}
+
+static void CheckExtendTimer(nlTestSuite * inSuite, void * aContext)
+{
+    uint32_t timeout_50ms  = 50;
+    uint32_t timeout_100ms = 100;
+
+    TestContext & lContext = *static_cast<TestContext *>(aContext);
+    Layer & lSys           = *lContext.mLayer;
+
+    sStartTimerHandled = false;
+
+    Error err = CHIP_SYSTEM_NO_ERROR;
+    struct timeval sleepTime;
+    sleepTime.tv_sec  = 0;
+    sleepTime.tv_usec = 1000; // 1 ms tick
+
+    err = lSys.StartTimer(timeout_100ms, HandleStartTimer, aContext);
+    NL_TEST_ASSERT(lContext.mTestSuite, err == CHIP_SYSTEM_NO_ERROR);
+
+    // Sleep for 20ms before extending the timer
+    for (int i = 0; i < 20; i++)
+    {
+        ServiceEvents(lSys, sleepTime);
+    }
+
+    err = lSys.ExtendTimer(timeout_50ms, HandleStartTimer, aContext);
+    NL_TEST_ASSERT(lContext.mTestSuite, err == CHIP_SYSTEM_NO_ERROR);
+
+    while (!sStartTimerHandled)
+    {
+        ServiceEvents(lSys, sleepTime);
+    }
+
+    NL_TEST_ASSERT(lContext.mTestSuite, sStartTimerHandled == true);
```

Comment:
Shouldn't we have a test confirming that this fails with NULL or mismatched context?

Reply:
Added